### PR TITLE
Define index/vertex buffer OOB behavior

### PR DESCRIPTION
**Pull request by kainino0x**
(Note: Can review with whitespace changes ignored.)

I believe these are the semantics agreed upon as recorded in #955.
Takes advantage of the robustness behavior defined by WGSL in #1722.

***
<a href="https://pr-preview.s3.amazonaws.com/kainino0x/gpuweb/pull/1835.html" title="Last updated on Jul 19, 2021, 11:55 PM UTC (000362b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1835/150002c...kainino0x:000362b.html" title="Last updated on Jul 19, 2021, 11:55 PM UTC (000362b)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -8403,50 +8404,58 @@ a list of vertices to process for each instance.
...
-            1. initialize the |vertexIndexList| with |drawCall|.vertexCount integers.
-            1. assign the |vertexIndexList| item |i| to be |drawCall|.firstVertex + |i|
-        1. Return |vertexIndexList|.
-
-        Note: in case of indirect draw calls, the `indexCount`, `vertexCount`,
-        and other properties of |drawCall| are read from the indirect buffer
-        instead of the draw command itself.
-
-        Issue: specify indirect commands better.
+    **Arguments:**
```
Comment (kvark) [inline spec/index.bs:8407]: Very unfortunate that indentation changes are mixed with the logic change here. I want to see what changed, and now github can't show me this

Comment (kainino0x) [reply spec/index.bs:8407]: Hiding whitespace changes on GitHub works fine for me:
https://github.com/gpuweb/gpuweb/pull/1835/files?diff=split&w=1

**Comment by Kangz** [comment]
Is this PR blocked on anything?

File: spec/index.bs [hunk 2]
```diff
@@ -8403,50 +8404,58 @@ a list of vertices to process for each instance.
...
+    **Returns:** list of integer indices.
+
+    1. Let |vertexIndexList| be an empty list of indices.
+    1. If |drawCall| is an indexed draw call:
+        1. Initialize the |vertexIndexList| with |drawCall|.indexCount integers.
+        1. For |i| in range 0 .. |drawCall|.indexCount (non-inclusive):
+            1. Let |relativeVertexIndex| be [$fetch index$](|i| + |drawCall|.`firstIndex`,
+                |state|.[=RenderState/indexBuffer=].`buffer`, |state|.[=RenderState/indexBuffer=].`offset`,
+                |state|.[=RenderState/indexBuffer=].`format`).
+            1. If |relativeVertexIndex| is `"out of bounds"`, stop and return the empty list.
```
Comment (kvark) [inline spec/index.bs:8420]: why are we comparing the relative vertex index, and not `|drawCall|.`baseVertex` + |relativeVertexIndex|` instead?

File: spec/index.bs [hunk 3]
```diff
@@ -8459,58 +8468,67 @@ clip space positions for [[#primitive-clipping]], as well as other data for the
...
+            1. Let |attributeOffset| be |vertexBufferOffset| +
+                |vertexElementIndex| * |vertexBufferLayout|.{{GPUVertexBufferLayout/arrayStride}} +
+                |attributeDesc|.{{GPUVertexAttribute/offset}}.
+            1. Load the attribute |data| of format |attributeDesc|.{{GPUVertexAttribute/format}}
+                from |vertexBuffer| starting at offset |attributeOffset|.
+                If this results in an out-of-bounds access, the resulting value is determined
+                according to WGSL's [=invalid memory reference=] behavior.
+            1. **Optionally (implementation-defined):**
+                If |attributeOffset| + sizeof(|attributeDesc|.{{GPUVertexAttribute/format}} &gt;
+                |vertexBuffer|.{{GPUBuffer/[[size]]}},
```
Comment (kvark) [inline spec/index.bs:8505]: we should be comparing against the active range specified to `setVertexBuffer`, not to the whole size

**Review by kvark — CHANGES_REQUESTED** [review]

File: spec/index.bs (hunk 2 above)
Comment (kainino0x) [reply spec/index.bs:8420]: This is not comparing the vertex index, it's checking whether the variable `relativeVertexIndex` has the special value `"out of bounds"` which is generated by the `fetch index` algorithm. Clarified.

File: spec/index.bs [hunk 4]
```diff
@@ -8668,50 +8669,59 @@ a list of vertices to process for each instance.
...
+    1. Let |elementByteSize| be defined by the |format|:
         <dl class="switch">
             : {{GPUIndexFormat/"uint16"}}
             :: 2
             : {{GPUIndexFormat/"uint32"}}
             :: 4
         </dl>
-        Interpret the data in |buffer| starting with |offset| + |i| * |stride|
-        of size |stride| bytes as an unsigned integer and return it.
+    1. If |offset| + |i + 1| &times; |elementByteSize| &gt; |buffer|.{{GPUBuffer/[[size]]}},
```
Comment (kvark) [inline spec/index.bs:8721]: we shouldn't be comparing this against the buffer size. Instead, we should be considering the binding size for the index buffer, i.e. stuff passed in the last argument to `setIndexBuffer()`, if any

File: spec/index.bs [hunk 5]
```diff
@@ -8668,50 +8669,59 @@ a list of vertices to process for each instance.
...
+    **Arguments:**
+    - |i|: Index of a vertex index to fetch.
+    - |buffer|: {{GPUBuffer}} containing index data.
+    - |offset|: Base offset into the |buffer|.
+    - |format|: {{GPUIndexFormat}} of the index.
 
-        Let |stride| be defined by the |format|:
+    **Returns:** unsigned integer or `"out of bounds"`
+
+    1. Let |elementByteSize| be defined by the |format|:
```
Comment (kvark) [inline spec/index.bs:8714]: maybe `indexSize`?

File: spec/index.bs [hunk 6]
```diff
@@ -8724,72 +8734,82 @@ clip space positions for [[#primitive-clipping]], as well as other data for the
...
+            1. Load the attribute |data| of format |attributeDesc|.{{GPUVertexAttribute/format}}
+                from |vertexBuffer| starting at offset |attributeOffset|.
+                The components are loaded in the order `x`, `y`, `z`, `w` from buffer memory.
+
+                If this results in an out-of-bounds access, the resulting value is determined
+                according to WGSL's [=invalid memory reference=] behavior.
+            1. **Optionally (implementation-defined):**
+                If |attributeOffset| + sizeof(|attributeDesc|.{{GPUVertexAttribute/format}}) &gt;
+                |vertexBufferOffset| + |vertexBufferBindingSize|,
+                [=list/empty=] |vertexIndexList| and stop, cancelling the draw call.
```
Comment (kvark) [inline spec/index.bs:8774]: Why do we need to `empty` it? What side effect does it give us versus just returning from this algorithm/function?

**Review by kvark — CHANGES_REQUESTED** [review]

File: spec/index.bs (hunk 6 above)
Comment (kainino0x) [reply spec/index.bs:8774]: Returning from here wouldn't prevent "Assemble primitives" and so on from trying to run. This should. But I guess we could just throw an exception...?

**Review by kvark — APPROVED** [review]

**Pull request merged**